[SPARK-50561][SQL] Improve type coercion and boundary checking for UNIFORM SQL function

dtenedor opened the pull request:
### What changes were proposed in this pull request?

This PR improve type coercion and boundary checking for UNIFORM SQL function.

@srielau found the following issues and wrote them down in SPARK-50561:
* TINYINT and BIGINT and DECIMAL types were not supported.
* No type coercion from floating-point numbers was implemented.
* No explicit error checking for negative numbers was implemented, resulting in weird stacktraces instead.

### Why are the changes needed?

This PR fixes the above problems to make the function work in more cases and produce better error messages when it fails.

For example:

```
SELECT uniform(cast(10 as decimal(10, 3)), cast(20 as decimal(10, 3)), 0.0D) AS result;
> 17.605

SELECT uniform(-20L, -10L, 0) AS result
> -12

SELECT uniform(0, cast(10 as tinyint), 0) AS result
> 7

SELECT uniform(0, cast(10 as bigint), 0) AS result
> 7
```

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds golden file based test coverage, and updates existing coverage.

### Was this patch authored or co-authored using generative AI tooling?

No.

dtenedor committed:
commit

commit

dtenedor committed:
commit

dtenedor commented:
cc @MaxGekk @HyukjinKwon @cloud-fan Greetings, this small PR is ready for a review at your convenience :)

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -229,16 +232,20 @@ case class Uniform(min: Expression, max: Expression, seedExpression: Expression,
 
   private def integer(t: DataType): Boolean = t match {
-    case _: ShortType | _: IntegerType | _: LongType => true
+    case _: ByteType | _: ShortType | _: IntegerType | _: LongType => true
```
Can you just check `IntegralType`?

cloud-fan replied:
+1

dtenedor replied:
I ended up just updating this to use the `ExpectsInputTypes` trait to simplify this code.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -49,6 +49,9 @@ trait RDG extends Expression with ExpressionWithRandomSeed {
     case e if e.dataType == LongType => e.eval().asInstanceOf[Long]
+    case e if e.dataType == FloatType => e.eval().asInstanceOf[Float].toLong
+    case e if e.dataType == DoubleType => e.eval().asInstanceOf[Double].toLong
+    case e if e.dataType.isInstanceOf[DecimalType] => e.eval().asInstanceOf[Decimal].toLong
```
truncation and overflow may happen here, how shall we deal it it?

cloud-fan replied:
BTW do we really expect users to use float/double/decimal as the random seed?

dtenedor replied:
I checked and the existing `RAND` and `RANDN` functions only accept `IntegerType` or `LongType` for the random seed (but positive, zero, and negative values are allowed). I updated this PR to be consistent with that.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -229,16 +232,20 @@ case class Uniform(min: Expression, max: Expression, seedExpression: Expression,
+        if Seq(first, second).forall(integer) => ByteType
       case (_, DoubleType) | (DoubleType, _) => DoubleType
       case (_, FloatType) | (FloatType, _) => FloatType
+      case (_, d: DecimalType) => d
```
shall we require the other side to be integral type?

cloud-fan replied:
or at least numeric type?

dtenedor replied:
Good idea, this is done.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -229,16 +232,20 @@ case class Uniform(min: Expression, max: Expression, seedExpression: Expression,
         if Seq(first, second).forall(integer) => IntegerType
       case (_, ShortType) | (ShortType, _)
```
One idea to simple it
```
case (left: IntegralType, right: IntegralType) =>
  if (UpCastRule.legalNumericPrecedence(left, right)) left else right
```

dtenedor replied:
Good idea, this is done.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -229,16 +232,20 @@ case class Uniform(min: Expression, max: Expression, seedExpression: Expression,
       case (_, DoubleType) | (DoubleType, _) => DoubleType
       case (_, FloatType) | (FloatType, _) => FloatType
+      case (_, d: DecimalType) => d
+      case (d: DecimalType, _) => d
```
if it's two decimals, shouldn't we pick the wider one instead of preferring the right one?

dtenedor replied:
Good idea, this is done.

dtenedor committed:
respond to code review comments

respond to code review comments

dtenedor reviewed (commented):
Thanks @MaxGekk and @cloud-fan for your reviews! Responded to your comments, please take another look.

dtenedor committed:
fix test

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -216,63 +216,48 @@ case class Uniform(min: Expression, max: Expression, seedExpression: Expression,
 
+  override def inputTypes: Seq[AbstractDataType] = {
+    val minMaxTypes =
+      TypeCollection(ByteType, ShortType, IntegerType, LongType, FloatType, DoubleType, DecimalType)
```
is it `NumericType`?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -290,19 +275,15 @@ case class Uniform(min: Expression, max: Expression, seedExpression: Expression,
     Uniform(newFirst, newSecond, newThird, hideSeed)
 
   override def replacement: Expression = {
-    if (Seq(min, max, seedExpression).exists(_.dataType == NullType)) {
```
why do we remove this branch?

dtenedor committed:
respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:
```diff
@@ -298,8 +281,8 @@ case class Uniform(min: Expression, max: Expression, seedExpression: Expression,
           Subtract(
-            cast(max, DoubleType),
-            cast(min, DoubleType)),
+            cast(If(IsNull(max), Literal(null, max.dataType), max), DoubleType),
```
hmm is this necessary? Cast is null-propagating, so null input results to null output already.

cloud-fan commented on sql/core/src/test/resources/sql-tests/inputs/random.sql:
```diff
@@ -22,21 +22,44 @@ SELECT uniform(0, 1, 0) AS result;
 SELECT uniform(0L, 10L, 0) AS result;
 SELECT uniform(0, 10L, 0) AS result;
+SELECT uniform(0, cast(10 as tinyint), 0) AS result;
+SELECT uniform(0, cast(10 as bigint), 0) AS result;
```
This test is the same as `SELECT uniform(0, 10L, 0) AS result;`. Shall we test a different type like `smallint`?

dtenedor replied:
Good idea, updated.

dtenedor committed:
respond to code review comments

cloud-fan reviewed (approved):
LGTM. There are few test failures need to be fixed.

dtenedor committed:
fix test

dtenedor commented:
Thanks, fixed the relevant test failures. The last failure look flaky/unrelated.

cloud-fan commented:
I think the CI has an issue now
```
Oh no! 💥 💔 💥 The required version `23.12.1` does not match the running version `23.9.1`!
Traceback (most recent call last):
Start checking the generated codes in pyspark-connect.
  File "/home/runner/work/spark/spark/./dev/check-protos.py", line 84, in <module>
    check_protos(
  File "/home/runner/work/spark/spark/./dev/check-protos.py", line 49, in check_protos
    run_cmd(f"{SPARK_HOME}/dev/gen-protos.sh {module_name} {tmp}")
  File "/home/runner/work/spark/spark/./dev/check-protos.py", line 43, in run_cmd
    return subprocess.check_output(cmd.split(" ")).decode("utf-8")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.11/x64/lib/python3.11/subprocess.py", line 466, in check_output
RUN: /home/runner/work/spark/spark/dev/gen-protos.sh connect /tmp/check_connect__protosnrkbs0co
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.11/x64/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/home/runner/work/spark/spark/dev/gen-protos.sh', 'connect', '/tmp/check_connect__protosnrkbs0co']' returned non-zero exit status 1.
```

cc @HyukjinKwon @panbingkun

cloud-fan commented:
It's unrelated to this PR, I'm merging it to master, thanks!

panbingkun commented:
> I think the CI has an issue now
> 
> ```
> Oh no! 💥 💔 💥 The required version `23.12.1` does not match the running version `23.9.1`!
> Traceback (most recent call last):
> Start checking the generated codes in pyspark-connect.
>   File "/home/runner/work/spark/spark/./dev/check-protos.py", line 84, in <module>
>     check_protos(
>   File "/home/runner/work/spark/spark/./dev/check-protos.py", line 49, in check_protos
>     run_cmd(f"{SPARK_HOME}/dev/gen-protos.sh {module_name} {tmp}")
>   File "/home/runner/work/spark/spark/./dev/check-protos.py", line 43, in run_cmd
>     return subprocess.check_output(cmd.split(" ")).decode("utf-8")
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/opt/hostedtoolcache/Python/3.11.11/x64/lib/python3.11/subprocess.py", line 466, in check_output
> RUN: /home/runner/work/spark/spark/dev/gen-protos.sh connect /tmp/check_connect__protosnrkbs0co
>     return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/opt/hostedtoolcache/Python/3.11.11/x64/lib/python3.11/subprocess.py", line 571, in run
>     raise CalledProcessError(retcode, process.args,
> subprocess.CalledProcessError: Command '['/home/runner/work/spark/spark/dev/gen-protos.sh', 'connect', '/tmp/check_connect__protosnrkbs0co']' returned non-zero exit status 1.
> ```
> 
> cc @HyukjinKwon @panbingkun

<img width="909" alt="image" src="https://github.com/user-attachments/assets/feff1576-1e99-4cf2-b6ec-bde0f221de54" />
https://github.com/apache/spark/blob/f0f850dc21ecf928995736daf8b790082e099221/.github/workflows/build_and_test.yml#L757
<img width="630" alt="image" src="https://github.com/user-attachments/assets/d1576258-6e8d-4a37-aea4-75c96aaaff11" />


This issue is very weird. Master's black requires the installation of version `black==23.12.1`, while GA actually installs version `23.9.1`. Is it because of GitHub's cache?

cloud-fan commented:
very likely, as many other PRs can pass the CI.
